File: src/main/java/org/dxworks/githubminer/pagination/GithubPaginationUtils.java
package org.dxworks.githubminer.pagination;

import com.google.api.client.http.GenericUrl;
import com.google.api.client.http.HttpRequestInitializer;
import lombok.SneakyThrows;
import lombok.extern.slf4j.Slf4j;
import org.dxworks.githubminer.http.GithubHttpResponse;
import org.dxworks.githubminer.service.GithubApiService;
import org.dxworks.utils.java.rest.client.response.HttpResponse;

import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

@Slf4j
public class GithubPaginationUtils extends GithubApiService {

    public GithubPaginationUtils() {
    }

    public GithubPaginationUtils(String githubRootUrl) {
        super(githubRootUrl);
    }

    public GithubPaginationUtils(HttpRequestInitializer httpRequestInitializer) {
        super(httpRequestInitializer);
    }

    public GithubPaginationUtils(String githubRootUrl, HttpRequestInitializer httpRequestInitializer) {
        super(githubRootUrl, httpRequestInitializer);
    }

    @SneakyThrows
    public <T> List<T> getAllElements(GenericUrl firstUrl, Type type) {

        GithubHttpResponse httpResponse = (GithubHttpResponse) getHttpResponse(firstUrl);

        List<T> elements = new ArrayList<>();

        while (hasNextPageLink(httpResponse)) {
            elements.addAll((Collection<T>) httpResponse.parseAs(type));
            GenericUrl genericUrl = new GenericUrl(httpResponse.getPageLinks().getNext());
            httpResponse = (GithubHttpResponse) getHttpResponse(genericUrl);
        }

        if (httpResponse != null)
            elements.addAll((Collection<T>) httpResponse.parseAs(type));

        return elements;
    }

    private HttpResponse getHttpResponse(GenericUrl genericUrl) {
        log.info("Retrieving " + genericUrl.toString());
        return httpClient.get(genericUrl);
    }

    private boolean hasNextPageLink(GithubHttpResponse httpResponse) {
        return httpResponse != null && httpResponse.getPageLinks().getNext() != null;
    }
}

File: github-miner-core/src/main/kotlin/org/dxworks/githubminer/pagination/GithubPaginationUtils.kt
package org.dxworks.githubminer.pagination

import com.google.api.client.http.GenericUrl
import org.dxworks.githubminer.constants.ANONYMOUS
import org.dxworks.githubminer.constants.GITHUB_API_PATH
import org.dxworks.githubminer.http.GithubHttpResponse
import org.dxworks.githubminer.http.factory.GithubHttpClientFactory
import org.dxworks.githubminer.service.GithubApiService
import org.slf4j.LoggerFactory
import java.util.*

class GithubPaginationUtils(
    githubBasePath: String = GITHUB_API_PATH,
    githubTokens: List<String> = listOf(ANONYMOUS),
    clientFactory: GithubHttpClientFactory?
)
    : GithubApiService(githubBasePath, githubTokens, clientFactory) {

    fun <T> getAllElements(firstUrl: GenericUrl, responseParser: (GithubHttpResponse) -> List<T>): List<T> {
        var httpResponse = getHttpResponse(firstUrl)
        val elements: MutableList<T> = ArrayList()
        while (hasNextPageLink(httpResponse)) {
            elements.addAll(responseParser(httpResponse))
            val genericUrl = GenericUrl(httpResponse.pageLinks.next)
            httpResponse = getHttpResponse(genericUrl)
        }
        elements.addAll(responseParser(httpResponse))
        return elements
    }

    private fun getHttpResponse(genericUrl: GenericUrl): GithubHttpResponse {
        return GithubHttpResponse(httpClient.get(genericUrl))
    }

    private fun hasNextPageLink(httpResponse: GithubHttpResponse?): Boolean {
        return httpResponse != null && httpResponse.pageLinks.next?.isNotBlank() ?: false
    }

    companion object {
        private val log = LoggerFactory.getLogger(GithubPaginationUtils::class.java)
    }
}
